### PR TITLE
Make regex field optional

### PR DESCRIPTION
**Pull request by psarka**
This PR makes the regex field optional.

The issues of potentially [illegal path](https://github.com/ContinuumIO/intake/issues/265) or file name collisions are not affected by this PR.

Alleviates https://github.com/ContinuumIO/intake/issues/254. It does not fully resolve it, as if user does provide a regex which does not match, files are still dropped in the cwd instead of `cache_dir`.

### CODE REVIEW

**Comment by martindurant** [comment]
As it happens, this seems to make an illegal path on windows

**Comment by psarka** [comment]
oh dear :man_facepalming:

File: docs/source/catalog.rst [hunk 1]
```diff
@@ -287,6 +285,11 @@ The special value ``"catdir"`` implies that cached files will appear in the same
 catalog file in which the data source is defined, within a directory named "intake_cache". These will
 not appear in the cache usage reported by the CLI.
 
+Optionally, the cache section can have a ``regex`` attribute, that modifies the path of the cache on
+the disk. By default, the cache path is made by concatenating ``cache_dir``, dataset name, hash of
+the url, and the url itself (without the protocol). ``regex`` attribute allows to remove part of the
```
Comment (martindurant) [inline docs/source/catalog.rst]: Is there a better description of regex than this? It feels a little hard to understand - but indeed, the usage *is* hard to uderstand.

File: intake/source/cache.py [hunk 2]
```diff
@@ -91,11 +91,15 @@ def _ensure_cache_dir(self):
     def _munge_path(self, cache_subdir, urlpath):
         import re
 
-        regex = sanitize_path(self._spec['regex'])
         path = sanitize_path(urlpath)
 
+        if 'regex' in self._spec:
+            regex = r'%s' % sanitize_path(self._spec['regex'])
+        else:
+            regex = r'^'  # if no regex provided, use full path
```
Comment (martindurant) [inline intake/source/cache.py]: Interesting minimum-code solution!

File: docs/source/catalog.rst (hunk 1 above)
Comment (psarka) [reply docs/source/catalog.rst]: I was not able to come up with something clear and succinct, so decided to explicitly spell out what it does. 

What could be a good idea, is to warn user about cache files ending up in cwd if regex is not well specified. Where well specified regex to my understanding is one that matches part of the url that includes the beginning, but excludes the protocol. Should I add this?

File: intake/source/cache.py (hunk 2 above)
Comment (psarka) [reply intake/source/cache.py]: :)

File: docs/source/catalog.rst (hunk 1 above)
Comment (martindurant) [reply docs/source/catalog.rst]: > warn user about cache files ending up in cwd

Sorry, I thought that would no longer be the case, that they would instead end up in the cache directory, under the hash, but with a deeper path

**Comment by martindurant** [comment]
Fixes #254

File: docs/source/catalog.rst (hunk 1 above)
Comment (martindurant) [reply docs/source/catalog.rst]: > Should I add this?

but yes, I think exactly that wording would be good

**Comment by psarka** [comment]
Instead of documenting accidental writing to cwd I realized I can fix it easily.

Note that to be sure that file always ends up in the cache dir I removed the `return url_path if cache_path == path logic`. I have to admit I don't understand why would `url_path` be useful as `cache_path`, and the only failing unit test (which I removed too), did not reveal the usefulness.

**Comment by martindurant** [comment]
Fair enough. Could you please a test showing that files do not end up in the cwd for the failed regex case, but they do end up in the catalog directory path if so specified?

**Comment by psarka** [comment]
I think this should test that. Can you please take a look?

**Comment by martindurant** [comment]
Thanks for all your work and patience, @psarka

**Pull request merged**